BLSQ/iaso
# IA-2411 Allow multiple reference `Form` and reference `Instance`

Enable the possibility to have:

- multiple reference `Form`s per `OrgUnitType`
- multiple reference `Instance`s per `OrgUnit`

Related JIRA tickets :

- [IA-2411](https://bluesquare.atlassian.net/browse/IA-2411)
- [IA-2229](https://bluesquare.atlassian.net/browse/IA-2229) (Epic)

## Self proofreading checklist


## Changes

- back-end
    - models
        - `OrgUnitType`
            - replace `reference_form` (FK) by `reference_forms` (M2M)
            - migrate DB data from `reference_form` to `reference_forms`
        - `OrgUnit`
            - replace `reference_instance` (FK) by `reference_instances` (M2M)
            - migrate DB data from `reference_form` to `reference_forms`
         - `OrgUnitReferenceInstance`
             - used as through table
         - `Instance`
             - add the logic to flag and unflag an instance as reference
    - APIs
        - mobile 
            - adapt `FormSerializer` ([per request](https://bluesquare.atlassian.net/browse/IA-2411?focusedCommentId=59205))
                - new `reference_form_of_org_unit_types` field
            - adapt `OrgUnitTypeSerializerV1` (nothing exposed as requested by the mobile app development team)
               - remove all `reference*` fields (unused by the mobile)
            - adapt `MobileOrgUnitSerializer`
                - new `reference_instances` field
        - web
            - adapt `InstancesViewSet`
                - `list()`and `retrieve()`
            - adapt `OrgUnitTypeSerializerV2`
                - add `reference_forms` and `reference_forms_ids` fields
                - new `validate_reference_forms()` method
            - adapt `OrgUnitViewSet`
                - this is what the front-end uses to flag an instance as reference
                - new logic in `partial_update()` based on an explicit code to trigger a `flag` or `unflag` action
    - adapt tests
    - adapt fixtures
- front-end
    - allow multiple reference forms per `OrgUnitType`
    - allow multiple reference instances per `OrgUnit`
    - new logic to flag or unflag a `reference instance` (`instance` is called `submission` in the front-end)
    - new component to display and switch between multiple `reference instances`s
    - deleted [`OrgUnitActions.tsx`](https://github.com/BLSQ/iaso/blob/b66a3f0e0ffdc0f9bcada7e0b50f9c74f81c4fa4/hat/assets/js/apps/Iaso/domains/orgUnits/components/OrgUnitActions.tsx) and the `SpeedDialsAction`from the `Org unit` details page
        - replaced by a link to the `Instance` details page
    - adapt fixtures
    - adapt tests

## How to test

Front-end (watch the video below for help):

1. Go to `Org Units > Organisation unit type`
    - edit a `Org unit type`
    - set some `Reference form`s
1. Go to `Forms > Submissions`
    - filter by `Forms` (choose those flagged above)
    - `link` some `Forms` to one `Org unit` to create `Reference submission`s
-  Go to `Org unit` details
    - you can see one or more `Reference submission`s and switch between them
    - each `Reference submission` has a link to its details


## Video

https://github.com/BLSQ/iaso/assets/281139/a2ab7dad-6c53-4b46-a324-ca33d55ed204

## Notes

[IA-2411]: https://bluesquare.atlassian.net/browse/IA-2411?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[IA-2229]: https://bluesquare.atlassian.net/browse/IA-2229?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
